### PR TITLE
Fix duplicate definition of a remote model type

### PR DESCRIPTION
**Pull request by angfal**
#### Related issues

- connect to #81
- see also the previous attempt in #83

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by angfal** [comment]
It's a more correct fix. The `define` method is called when a model is attaching to a datasource. It will define an additional remote type and this type won't re-defined when relations is attached.

https://github.com/strongloop/loopback-datasource-juggler/blob/00cf01f901e5ff20f3a576f10f41779e70cf536f/lib/datasource.js#L633

**Comment by bajtos** [comment]
Thank you for the pull request, I'll try to take a look in the next few days.

File: test/models-define-type.test.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+const RelationTypes = relation.RelationTypes;
+
+describe('Models Define Type Tests', function() {
+  let serverApp, clientApp, remoteDs, defineObjectTypeSpy, ChildModel;
+
+  before('create remote datasource', () => {
+    serverApp = helper.createRestAppAndListen();
+    clientApp = loopback({localRegistry: true});
+    remoteDs = helper.createRemoteDataSource(clientApp, serverApp);
+    defineObjectTypeSpy = sinon.spy(remoteDs.connector.remotes,
```
Comment (bajtos) [inline test/models-define-type.test.js]: This creates a single spy shared by all tests below. As a result, the spy will record multiple invocations, depending on how many tests are being run (e.g. via `mocha -g` or `it.only()`). I am concerned that if something breaks, then unrelated test will start failing because they will see bad state from previous tests.

I am proposing to create a new spy before each test and restore the spied method after each test, or perhaps use your approach from https://github.com/strongloop/loopback-connector-remote/pull/83, where the spy is reset in a beforeEach hook.

File: lib/remote-connector.js [hunk 2]
```diff
@@ -101,6 +92,16 @@ RemoteConnector.prototype.resolve = function(Model) {
   });
 };
 
+RemoteConnector.prototype.resolve = function(Model) {
```
Comment (bajtos) [inline lib/remote-connector.js]: Please move this method back to lines 79-86, to avoid unnecessary changes in the git history?

**Comment by bajtos** [comment]
@slnode ok to test

File: lib/remote-connector.js (hunk 2 above)
Comment (angfal) [reply lib/remote-connector.js]: Hmmm. How it possible if I moved a part of this method to the method before it (to the `define` function)?

File: test/models-define-type.test.js (hunk 1 above)
Comment (angfal) [reply test/models-define-type.test.js]: Fixed

File: lib/remote-connector.js (hunk 2 above)
Comment (bajtos) [reply lib/remote-connector.js]: Ah, I see what you mean now. I extracted a new method `setupRemotingTypeFor` that allows us to keep the patch smaller - see 40adba0.

Now I am going to squash all commits into a single one, so that this pull request can be merged.

Comment (angfal) [reply lib/remote-connector.js]: Great! Thank you

**Comment by bajtos** [comment]
While cleaning up the git history, I have also fixed the tense of your commit message to imperative and added more details to commit message body. See http://loopback.io/doc/en/contrib/git-commit-messages.html

**Comment by bajtos** [comment]
Landed 🎉  Thank you for the contribution! ❤️
